### PR TITLE
[5.0] Batch Copy & Move tags

### PR DESCRIPTION
**Pull request by brianteeman**
Pull Request for Issue #23304 .

### Summary of Changes
Solves a four year old feature request ;)

Adds the ability to copy and move tags

https://user-images.githubusercontent.com/1296369/186410752-c120430b-35cf-40fe-ba92-0a713b3d827a.mp4

Its a new feature so its against the 4.3 branch

### Testing Instructions
It requires the tag alias bug fix from #38576
It contains an updated javascript so run npm build:js

See video

### CODE REVIEW

File: administrator/language/en-GB/com_tags.ini [hunk 1]
```diff
@@ -9,6 +9,7 @@ COM_TAGS_ALL_TAGS_DESCRIPTION_LABEL="Heading Description"
 COM_TAGS_ALL_TAGS_MEDIA_LABEL="Heading Image File"
 COM_TAGS_ANY="Any"
 COM_TAGS_BASIC_FIELDSET_LABEL="Options"
+COM_TAGS_BATCH_COPY_MOVE_CANNOT_CREATE="You do not have permission to copy or move a tag.."
```
Comment (Kostelano) [inline administrator/language/en-GB/com_tags.ini]: Two dots

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 2]
```diff
@@ -406,4 +417,280 @@ protected function generateNewTitle($parentId, $alias, $title)
...
+     * @return  mixed  An array of new IDs on success, boolean false on failure.
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function batchCopy($value, $pks, $contexts)
+    {
+        $table     = $this->getTable();
+        $db        = $this->getDatabase();
+        $query     = $db->getQuery(true);
+        $newIds    = array();
```
Comment (heelc29) [inline administrator/components/com_tags/src/Model/TagModel.php:437]: short array syntax
```suggestion
        $newIds    = [];
```

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 3]
```diff
@@ -406,4 +417,280 @@ protected function generateNewTitle($parentId, $alias, $title)
...
+
+        // Check that user has create permission for tags
+        if (!Factory::getApplication()->getIdentity()->authorise('core.create', 'com_tags')) {
+            $this->setError(Text::_('COM_TAGS_BATCH_COPY_MOVE_CANNOT_CREATE'));
+
+            return false;
+        }
+
+        // We need to log the parent ID
+        $parents = array();
```
Comment (heelc29) [inline administrator/components/com_tags/src/Model/TagModel.php:473]: short array syntax
```suggestion
        $parents = [];
```

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 4]
```diff
@@ -406,4 +417,280 @@ protected function generateNewTitle($parentId, $alias, $title)
...
+        if (empty($parentId)) {
+            if (!$parentId = $table->getRootId()) {
+                $this->setError($table->getError());
+
+                return false;
+            }
+        }
+
+        // Check that user has create permission for tags
+        if (!Factory::getApplication()->getIdentity()->authorise('core.create', 'com_tags')) {
```
Comment (heelc29) [inline administrator/components/com_tags/src/Model/TagModel.php]: See #37498
```suggestion
        if (!$this->getCurrentUser()->authorise('core.create', 'com_tags')) {
```

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 5]
```diff
@@ -406,4 +417,280 @@ protected function generateNewTitle($parentId, $alias, $title)
...
+            // Make a copy of the old ID and Parent ID
+            $oldId = $table->id;
+            $oldParentId = $table->parent_id;
+
+            // Reset the id because we are making a copy.
+            $table->id = 0;
+
+            // If we a copying children, the Old ID will turn up in the parents list
+            // otherwise it's a new top level item
+            $table->parent_id = isset($parents[$oldParentId]) ? $parents[$oldParentId] : $parentId;
```
Comment (heelc29) [inline administrator/components/com_tags/src/Model/TagModel.php:540]: null coalescing
```suggestion
            $table->parent_id = $parents[$oldParentId] ?? $parentId;
```

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 6]
```diff
@@ -406,4 +417,280 @@ protected function generateNewTitle($parentId, $alias, $title)
...
+            $table->setLocation($table->parent_id, 'last-child');
+
+            // @todo: Deal with ordering?
+            // $table->ordering = 1;
+            $table->level = null;
+            $table->lft   = null;
+            $table->rgt   = null;
+
+            // Alter the title & alias
+            list($title, $alias) = $this->generateNewTitle($table->parent_id, $table->alias, $table->title);
```
Comment (heelc29) [inline administrator/components/com_tags/src/Model/TagModel.php:553]: short array syntax
```suggestion
            [$title, $alias] = $this->generateNewTitle($table->parent_id, $table->alias, $table->title);
```

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 7]
```diff
@@ -406,4 +417,280 @@ protected function generateNewTitle($parentId, $alias, $title)
...
+     * @return  mixed  An array of new IDs on success, boolean false on failure.
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function batchMove($value, $pks, $contexts)
+    {
+        $table     = $this->getTable();
+        $db        = $this->getDatabase();
+        $query     = $db->getQuery(true);
+        $newIds    = array();
```
Comment (heelc29) [inline administrator/components/com_tags/src/Model/TagModel.php:618]: short array syntax
```suggestion
        $newIds    = [];
```

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 8]
```diff
@@ -406,4 +417,280 @@ protected function generateNewTitle($parentId, $alias, $title)
...
+                } else {
+                    // Non-fatal error
+                    $this->setError(Text::_('JGLOBAL_BATCH_MOVE_PARENT_NOT_FOUND'));
+                    $parentId = 0;
+                }
+            }
+        }
+
+        // Check that user has create and edit permission for menus
+        if (!Factory::getApplication()->getIdentity()->authorise('core.create', 'com_tags')) {
```
Comment (heelc29) [inline administrator/components/com_tags/src/Model/TagModel.php]: See #37498
```suggestion
        if (!$this->getCurrentUser()->authorise('core.create', 'com_tags')) {
```

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 9]
```diff
@@ -406,4 +417,280 @@ protected function generateNewTitle($parentId, $alias, $title)
...
+            $this->setError(Text::_('COM_TAGS_BATCH_COPY_MOVE_CANNOT_CREATE'));
+
+            return false;
+        }
+
+        // We are going to store all the children and just moved the menutype
+        $children = array();
```
Comment (heelc29) [inline administrator/components/com_tags/src/Model/TagModel.php:645]: short array syntax
```suggestion
        $children = [];
```

File: build/media_source/layouts/js/joomla/html/batch/batch-copymove.es6.js [hunk 10]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable linebreak-style */
```
Comment (heelc29) [inline build/media_source/layouts/js/joomla/html/batch/batch-copymove.es6.js]: Is this required?

**Comment by brianteeman** [comment]
I'm not making the requested changes. It makes no sense to me to have two different code styles for identical code. When phpcs finally detects these things then it can automatically do it.

**Comment by heelc29** [comment]
> I'm not making the requested changes. It makes no sense to me to have two different code styles for identical code. When phpcs finally detects these things then it can automatically do it.

Ok. I'll delete the comments for _short array syntax_.
But what about the changes with `getCurrentUser`?. Here the other components have been adjusted in 4.3-dev.

**Comment by brianteeman** [comment]
> But what about the changes with getCurrentUser?. Here the other components have been adjusted in 4.3-dev.

Yes that one makes sense and I have accepted the change

File: build/media_source/layouts/js/joomla/html/batch/batch-copymove.es6.js (hunk 10 above)
Comment (brianteeman) [reply build/media_source/layouts/js/joomla/html/batch/batch-copymove.es6.js]: it was added by my ide and I didnt spot it. Removing it now and we'll see if it passes the cs

**Comment by brianteeman** [comment]
Updating branch to retrigger drone

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 11]
```diff
@@ -13,11 +13,13 @@
 use Joomla\CMS\Component\ComponentHelper;
 use Joomla\CMS\Date\Date;
 use Joomla\CMS\Factory;
+use Joomla\CMS\Language\Text;
 use Joomla\CMS\MVC\Model\AdminModel;
 use Joomla\CMS\Plugin\PluginHelper;
 use Joomla\CMS\Versioning\VersionableModelTrait;
 use Joomla\Registry\Registry;
 use Joomla\String\StringHelper;
+use Joomla\Database\ParameterType;
```
Comment (Quy) [inline administrator/components/com_tags/src/Model/TagModel.php]:
```suggestion
use Joomla\Database\ParameterType;
use Joomla\Registry\Registry;
use Joomla\String\StringHelper;
```

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 12]
```diff
@@ -44,6 +46,15 @@ class TagModel extends AdminModel
      */
     public $typeAlias = 'com_tags.tag';
 
+     /**
```
Comment (Quy) [inline administrator/components/com_tags/src/Model/TagModel.php]:
```suggestion
    /**
```

**Comment by brianteeman** [comment]
Updating branch to restart the broken drone

**Comment by brianteeman** [comment]
update array code style

**Comment by ceford** [comment]
No video so can't test.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38583">issues.joomla.org/tracker/joomla-cms/38583</a>.</sub>

**Comment by beni71** [comment]
I tried to download the prebuilt package by clicking Details (above) but it results in a "Not found" error.

**Comment by HLeithner** [comment]
PR is too old, New link will be created shortly

**Comment by beni71** [comment]
I tried to test it by installing the full package from download > Details link. But in the batch popup the new field is missing. (I'm testing with super user rights). When I debug it, the following layout

`<?php echo LayoutHelper::render('joomla.html.batch.tagcopymove', []); ?>`

could not be found (does not exist). Shouldn't it be here: /layouts/joomla/html/batch/tagcopymove.php

**Comment by beni71** [comment]
I have tested this item :red_circle: unsuccessfully on d62ae9aa452a391353ad3ba9b25d32c13f39fa7d<br /><br />See my comment above.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38583">issues.joomla.org/tracker/joomla-cms/38583</a>.</sub>

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 13]
```diff
@@ -410,4 +421,276 @@ protected function generateNewTitle($parentId, $alias, $title)
...
+     *
+     * @return  mixed  An array of new IDs on success, boolean false on failure.
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    protected function batchMove($value, $pks, $contexts)
+    {
+        $table     = $this->getTable();
+        $db        = $this->getDatabase();
+        $query     = $db->getQuery(true);
```
Comment (beni71) [inline administrator/components/com_tags/src/Model/TagModel.php:619]: `$query` is not used. Can be removed, same for `$db`.

File: administrator/components/com_tags/src/Model/TagModel.php [hunk 14]
```diff
@@ -410,4 +421,276 @@ protected function generateNewTitle($parentId, $alias, $title)
...
+        $newIds    = [];
```
Comment (beni71) [inline administrator/components/com_tags/src/Model/TagModel.php:620]: The code block of the function states that it returns an array on success, but it returns a bool. I don't know what is correct, but if bool is the way to go then `$newIds` is never used.

**Comment by brianteeman** [comment]
feel free to submit a pr to fix anything you see to my branch. I wrote this 13 months ago for joomla 4.3 and I dont really have the time to dig back into the code myself to look at the changes for this

**Comment by brianteeman** [comment]
closed see #41927

**Pull request closed without merging**